pf/internal/plugin/provider_server.go

```diff
@@ -16,702 +16,80 @@ package plugin
-
-type forwardServer struct {
-	plugin.UnimplementedProvider
-
-	parameterize func(context.Context, plugin.ParameterizeRequest) (plugin.ParameterizeResponse, error)
-}
-
-func (p forwardServer) Parameterize(
-	ctx context.Context, req plugin.ParameterizeRequest,
-) (plugin.ParameterizeResponse, error) {
-	return p.parameterize(ctx, req)
-}
-
-func (p *providerServer) Parameterize(
-	ctx context.Context, req *pulumirpc.ParameterizeRequest,
-) (*pulumirpc.ParameterizeResponse, error) {
-	return plugin.NewProviderServer(&forwardServer{
-		parameterize: p.provider.ParameterizeWithContext,
-	}).Parameterize(ctx, req)
-}
-
-func (p *providerServer) GetSchema(ctx context.Context,
-	req *pulumirpc.GetSchemaRequest,
-) (*pulumirpc.GetSchemaResponse, error) {
-	var subpackageVersion *semver.Version
-	if v := req.GetSubpackageVersion(); v != "" {
-		ver, err := semver.Parse(v)
-		if err != nil {
-			return nil, fmt.Errorf("invalid SubpackageVersion: %w", err)
-		}
-		subpackageVersion = &ver
-	}
-	schema, err := p.provider.GetSchemaWithContext(ctx, plugin.GetSchemaRequest{
-		Version:           int(req.GetVersion()),
-		SubpackageName:    req.SubpackageName,
-		SubpackageVersion: subpackageVersion,
-	})
-	if err != nil {
-		return nil, err
-	}
-	return &pulumirpc.GetSchemaResponse{Schema: string(schema)}, nil
-}
-
-func (p *providerServer) GetPluginInfo(ctx context.Context, req *pbempty.Empty) (*pulumirpc.PluginInfo, error) {
-	info, err := p.provider.GetPluginInfoWithContext(ctx)
-	if err != nil {
-		return nil, err
-	}
-	return &pulumirpc.PluginInfo{Version: info.Version.String()}, nil
-}
-
-func (p *providerServer) Attach(ctx context.Context, req *pulumirpc.PluginAttach) (*pbempty.Empty, error) {
-	// NewProviderServer should take a GrpcProvider instead of Provider, but that's a breaking change
-	// so for now we type test here
-	if grpcProvider, ok := p.provider.(interface {
-		Attach(address string) error
-	}); ok {
-		err := grpcProvider.Attach(req.GetAddress())
-		if err != nil {
-			return nil, err
-		}
-		return &pbempty.Empty{}, nil
-	}
-	// Else report this is unsupported
-	return nil, status.Error(codes.Unimplemented, "Attach is not yet implemented")
+	p := &providerServer{
+		ResourceProviderServer: plugin.NewProviderServer(
+			providerThunk{
+				configencoding.New(configEncoding, NewProvider(provider)),
+			},
+		),
+	}
+
+	// This is a hack to work around limitations in the Pulumi protocol handshake:
+	//
+	// We want to be able to return secret values from CheckConfig, but we don't know
+	// if the engine accepts secrets until Configure is called (after
+	// CheckConfig). [plugin.NewProviderServer] defaults to not sending secrets, which
+	// is safe but breaks existing behavior.
+	//
+	// We send an artificial Configure call to tell the
+	// [pulumirpc.ResourceProviderServer] that it should accept
+	// secrets. [providerThunk] acts as a thunk, preventing the underlying
+	// [ProviderWithContext] from witnessing the extra Configure call.
```

Comment:
What is the existing behavior? Are bridged providers assuming that the engine accepts secrets, even though the older versions of the engine do not?

Reply:
Actually, curious, can we fix the protocol and send this information (whether the engine supports secrets or not) as part of CheckRequest.

Reply:
If we are going to change the protocol:

We shouldn't do it as part of `CheckRequest`, we should have the engine send a `HandshakeRequest` that handles protocol level negotiation. Let's leave `CheckConfig` and `Configure` to be provider level options.